File: 0146.LRU_cache/solution.py
class DoublyLinkedList:
    """ 双向链表 """

    def __init__(self, key=0, val=0):
        self.pre = None
        self.next = None
        self.key = key
        self.val = val


class LRUCache:
    """ LRU (双向链表结合哈希表)"""

    def __init__(self, capacity: int):
        self.cap = capacity
        self.cache = {}
        self.head = DoublyLinkedList()
        self.tail = DoublyLinkedList()
        self.head.next = self.tail
        self.tail.pre = self.head
        self.size = 0

    def get(self, key: int) -> int:
        if key in self.cache:
            node = self.cache[key]
            self.remove_node(node)
            self.move_to_head(node)
            return node.val
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            node = self.cache[key]
            node.val = value
            self.remove_node(node)
            self.move_to_head(node)
        else:
            node = DoublyLinkedList(key, value)
            self.cache[key] = node
            self.move_to_head(node)
            self.size += 1
            if self.size > self.cap:
                node = self.tail.pre
                self.remove_node(node)
                self.cache.pop(node.key)
                self.size -= 1

    def remove_node(self, node):
        node.pre.next = node.next
        node.next.pre = node.pre

    def move_to_head(self, node):
        node.pre = self.head
        node.next = self.head.next
        self.head.next.pre = node
        self.head.next = node

File: 0146.LRU_cache/solution.go
package leetcode0146

// DoublyLinkedList : 双向链表
type DoublyLinkedList struct {
	prev, next *DoublyLinkedList
	key, value int
}

// LRUCache : LRU cache
type LRUCache struct {
	cap, size  int
	cache      map[int]*DoublyLinkedList
	head, tail *DoublyLinkedList
}

// Constructor : init a LRUCache with capacity
func Constructor(capacity int) LRUCache {
	cache := LRUCache{
		cap: capacity,
		head: &DoublyLinkedList{key: 0, value: 0},
		tail: &DoublyLinkedList{key: 0, value: 0},
		cache: map[int]*DoublyLinkedList{},
	}
	cache.head.next = cache.tail
	cache.tail.prev = cache.head
	return cache
}

// Get : LRUCache get a node
func (c *LRUCache) Get(key int) int {
	if _, ok := c.cache[key]; !ok {
		return -1
	}
	node := c.cache[key]
	// remove node
	node.prev.next = node.next
	node.next.prev = node.prev

	// add to head
	node.prev = c.head
	node.next = c.head.next
	c.head.next.prev = node
	c.head.next = node
	return node.value
}

// Put : LRUCache put a node
func (c *LRUCache) Put(key int, value int) {
	if _, ok := c.cache[key]; !ok {
		node := &DoublyLinkedList{key: key, value: value}
		c.cache[key] = node
		// add to head
		node.prev = c.head
		node.next = c.head.next
		c.head.next.prev = node
		c.head.next = node
		c.size++
		if c.size > c.cap {
			// remove the tail node
			node := c.tail.prev
			node.prev.next = node.next
			node.next.prev = node.prev
			delete(c.cache, node.key)
			c.size--
		}
	} else {
		node := c.cache[key]
		// change value
		node.value = value
		// remove node
		node.prev.next = node.next
		node.next.prev = node.prev
		// add to head
		node.prev = c.head
		node.next = c.head.next
		c.head.next.prev = node
		c.head.next = node
	}
}
